Command arguments

Goldorion opened the pull request:
This PR implements a new Blockly Panel inside the command mod element allowing users to fully add and use command arguments. I decided to use a Blockly for this feature instead of something similar to loot tables because Blockly is much easier to understand what we are making when we use several arguments and sub-arguments.

This way commands like `/setfire blocks 2` and `/setfire entities @e 2` will be possible.

Command argument and procedure blocks are added via #1183, a the exception of one cmd arg block for new tests.

KlemenDEV commented:
Are you sure Blockly is needed here, and not list like eg. biome spawn entries? I would avoid blockly when not really needed.

Goldorion commented:
> 
> 
> Are you sure Blockly is needed here, and not list like eg. biome spawn entries? I would avoid blockly when not really needed.

I tried with that at the beginning, but as soon as you have a sub-argument inside an argument, you don't understand what is what.

Goldorion commented:
I'm closing the PR for the moment because I don't know if I will be able to finish it before the 3rd snapshot.

Goldorion commented:
Everything I wanted and needed to add has been added. Please tests with old workspaces, creating new ones and try to make commands with several arguments and argument paths. I will continue my tests, but I also need external tests. Concerning the templates, I will try to make some before we merge the PR. If you have some good templates, you can add them obviously. :)

Goldorion commented:
I kept 5 argument blocks because Travis needs them to don't have an error. All other blocks have been moved to #1183.

KlemenDEV commented:
> I kept 5 argument blocks because Travis needs them to don't have an error. All other blocks have been moved to #1183.

You should write tests to test all blocks available and not hardcode tests for a few blocks only, or tests may miss bugs when eg. updating generator

Goldorion commented:
> 
> 
> > I kept 5 argument blocks because Travis needs them to don't have an error. All other blocks have been moved to #1183.
> 
> You should write tests to test all blocks available and not hardcode tests for a few blocks only, or tests may miss bugs when eg. updating generator

No problem, but the test for the AI has hard coded task blocks.

KlemenDEV commented:
> > > I kept 5 argument blocks because Travis needs them to don't have an error. All other blocks have been moved to #1183.
> > 
> > 
> > You should write tests to test all blocks available and not hardcode tests for a few blocks only, or tests may miss bugs when eg. updating generator
> 
> No problem, but the tests for the AI and the advancement trigger have hard coded blocks.

I am aware of that, I did not have time to port them yet, but at least new features should have them. I do plan to add them for AI and advancements too, but did not get the time yet :D

KlemenDEV commented:
> I know with the way I implemented the thing, it won't work properly every time. However, I don't know how we can know there are no blocks under the starting block.

Does this still apply?

Goldorion commented:
> > I know with the way I implemented the thing, it won't work properly every time. However, I don't know how we can know there are no blocks under the starting block.
> 
> Does this still apply?

I didn’t change the code of the preview, so yes surely.

KlemenDEV commented:
> > > I know with the way I implemented the thing, it won't work properly every time. However, I don't know how we can know there are no blocks under the starting block.
> > 
> > 
> > Does this still apply?
> 
> I didn’t change the code of the preview, so yes surely.

This part needs to be figured out because the block editor itself should be as "foolproof" as possible

Goldorion commented:
> 
> 
> > > > I know with the way I implemented the thing, it won't work properly every time. However, I don't know how we can know there are no blocks under the starting block.
> > > 
> > > 
> > > Does this still apply?
> > 
> > 
> > I didn’t change the code of the preview, so yes surely.
> 
> This part needs to be figured out because the block editor itself should be as "foolproof" as possible

I don't know how to make a better image mod element preview. If you want to have the right mod element preview each time, you will have to change it yourself because I've no idea. In addition, it's not important as there will always be the old (current) preview as the default one.

KlemenDEV commented:
> I know with the way I implemented the thing, it won't work properly every time. 

So this statement was just about previews, not about some bigger issue? I understood the command editor itself wont work properly every time :D

Goldorion commented:
> > I know with the way I implemented the thing, it won't work properly every time.
> 
> So this statement was just about previews, not about some bigger issue? I understood the command editor itself wont work properly every time :D

You never saw the small `*`? xD

KlemenDEV commented:
> > > I know with the way I implemented the thing, it won't work properly every time.
> > 
> > 
> > So this statement was just about previews, not about some bigger issue? I understood the command editor itself wont work properly every time :D
> 
> You never saw the small `*`? xD

Oh I see it now :D

Goldorion commented:
I will fix the tests tomorrow. I think I know what I have to add to fix them.

KlemenDEV commented on plugins/generator-1.15.2/forge-1.15.2/mappings/types.yaml:
```diff
@@ -8,4 +8,5 @@ string: String
-dimensiontype: int
\ No newline at end of file
+dimensiontype: int
+ctx: CommandContext<CommandSource>
```
Call this commandcontext or cmdcontext so it is more descriptive. Ctx could be many things

KlemenDEV commented:
Merge conflicts need fixing

KlemenDEV commented:
This PR has conflicts that need attention

Goldorion commented:
I will try to fix the tests tomorrow.

Max094Reikeb commented:
> I will try to fix the tests tomorrow.

Lol it's been a week

Goldorion commented:
> 
> 
> > I will try to fix the tests tomorrow.
> 
> Lol it's been a week

Heatwave, school, boring things XD It's the next PR I want to finish.

KlemenDEV commented on plugins/mcreator-localization/help/default/command/when_executed.md:
```diff
@@ -1,3 +1,5 @@
\ No newline at end of file
+Commands usually run on the server-side only.
+
+Note: You should only use this procedure selector if you do not need parameters.
```
Consider removing procedure selector altogether in favor of new builder. One can use call procedure procedure block to call procedure from there. Also prepare migration path for existing workspaces for this.

KlemenDEV commented on src/main/java/net/mcreator/ui/blockly/BlocklyJavascriptBridge.java:
```diff
@@ -138,6 +139,24 @@
 		callback.call("callback", StringUtils.join(retval, ','));
 	}
 
+	@SuppressWarnings("unused") public void openArgExecuteProcedureEditor(String data, JSObject callback) {
```
You should use existing "call procedure block" for this. Add this block to your toolbox and drop setting execute procedure with this so we don't duplicate existing functionality if possible.

Goldorion replied:
Simple questio. How do you want to use all dependencies provided by the command context (including arguments) without adding a new block or a new interface like you made for AI conditons?

KlemenDEV replied:
Call procedure block automatically loads all dependencies required by the calling procedure so if this procedure requires arguments, they will be added to the dependency list

Goldorion commented:
I added a new category named "Actions" because this new system will allow us to add new things like the `suggests` method for arguments. Currently, it only has the call procedure block, but in the future, blocks doing something (like the example of the previous example) will go in this category. Command parameters now only require 1 block with a statement. I'll change https://github.com/MCreator/MCreator/pull/1183 as soon as this PR will be merged (or ready to be merged).

KlemenDEV commented:
> I added a new category named "Actions" because this new system will allow us to add new things like the `suggests` method for arguments. Currently, it only has the call procedure block, but in the future, blocks doing something (like the example of the previous example) will go in this category. Command parameters now only require 1 block with a statement. I'll change #1183 as soon as this PR will be merged (or ready to be merged).

Most procedures are actions - they do something, so I would suggest looking for alternative way of sorting these blocks

Goldorion commented:
6 months later, this PR feels abandoned.

KlemenDEV commented:
This PR has some conflicts to fix

Goldorion commented:
> This PR has some conflicts to fix

I fixed conflicts and I also added code templates for Forge 1.18.1.

Goldorion commented:
Marking as draft until changes of #2239 are applied.

KlemenDEV commented on src/test/java/net/mcreator/integration/generator/GeneratorsTest.java:
```diff
@@ -151,6 +152,19 @@
+				LOG.info("[" + generator + "] ----- Testing building after procedure tests");
+				GTBuild.runTest(LOG, generator, workspace);
+
+				LOG.info("[" + generator + "] ----- Testing command argument blocks");
```
Now, single build is run so do not build here

KlemenDEV replied:
You marked this as resolved, but it was not

Goldorion commented:
@KlemenDEV you should the stale bot for PRs. It useless, we only remove the label to be readded 2 weeks later and it spams comments.

KlemenDEV commented:
I can remove it for you :)

Goldorion commented:
It is its birthday Sunday! Who's making the cake?

Max094Reikeb commented:
You should resolve merging conflicts...

KlemenDEV commented on src/main/java/net/mcreator/element/converter/fv26/CommandArgumentBlockFixer.java:
```diff
@@ -0,0 +1,40 @@
+	@Override
+	public GeneratableElement convert(Workspace workspace, GeneratableElement input, JsonElement jsonElementInput) {
+		Command command = (Command) input;
+		command.argsxml = "<xml><block type=\"args_start\" deletable=\"false\" x=\"40\" y=\"40\"></block></xml>";
```
You should also by default convert existing commands to use procedure call block to call the procedure they used before

Goldorion replied:
I am not against the idea. However, every command using the old command parameter getter block (so the old dependency too) will not have a working command command in the game, but they will not be able to notice until they test it as it will not produce errors inside MCreator. The best thing would be to convert everything, procedures and add blocks inside the new Blockly editor, but it may not be possible with the current way the converter works. We should at least warns the user its commands will not work anymore.

KlemenDEV replied:
So the new system does not support calling procedure on command as before?

Goldorion replied:
What I meant is this new system does not support the old command dependency where you need to write the number of the argument to get it as a string. Bow, you need the name of the argument you wrote inside the command argument block you placed inside the Blockly panel. This system still uses a procedure to execute actions.

KlemenDEV replied:
I see. I would say make a converter that converts procedures that do not use command arguments. For complex procedures, we will add release note to redesign those commands.

It would also be worth considering adding command argument "anything" to support sort of functionality there is right now, possibly

Goldorion replied:
> It would also be worth considering adding command argument "anything" to support sort of functionality there is right now, possibly

I will add it in the other PR after you merge this first PR into a temporary branch.

KlemenDEV replied:
It may be good to have it in this one, so the converter is valid, or at least make converter convert all but anything parameter here

KlemenDEV commented on plugins/mcreator-localization/lang/texts.properties:
```diff
@@ -2587,6 +2603,7 @@ dialog.generator_selector.coverage.procedure_blocks=Procedure blocks
 dialog.generator_selector.coverage.ai_tasks=AI tasks/goals
 dialog.generator_selector.coverage.global_triggers=Global triggers
 dialog.generator_selector.coverage.advancement_triggers=Advancement triggers
+dialog.generator_selector.coverage.cmd_args=Command parameters
```
I would call that command arguments

KlemenDEV reviewed (changes requested):
Also this was not tested at all:

![image](https://user-images.githubusercontent.com/16374228/155517087-0612eb3b-b5c7-489d-9576-f3f5417c09cc.png)

Goldorion commented:
> Also this was not tested at all:
> 
> ![image](https://user-images.githubusercontent.com/16374228/155517087-0612eb3b-b5c7-489d-9576-f3f5417c09cc.png)

I’m pretty sure I tested it before, but when I updated to the new Blockly system, I forgot to re test this part (or I simply forgot to add texts).

Goldorion commented:
@KlemenDEV I just added the "anything" block you suggested. It's named `old_command`. When you use this block, the command will get a new argument with the `execute` part coming just after. As this `execute` is linked to the argument, it will allow users to use it at the same time as using other argument types. I also modified the converter to add this block only if a procedure was already selected. if no procedure was selected, I simply add the start block.

Goldorion commented:
> A merge conflict needs to be resolved

Done

KlemenDEV reviewed (commented):
1. How will additional procedures know what context data is provided and what not, eg. I add a block position parameter called "pos", how will validation make sure the user is using argument actually existing?
2. I would keep procedure editor and two simple parameters in a single page

Goldorion commented:
> 1. How will additional procedures know what context data is provided and what not, eg. I add a block position parameter called "pos", how will validation make sure the user is using argument actually existing?
> 
>     2. I would keep procedure editor and two simple parameters in a single page

1. If you are talking about getter procedure blocks, this is a similar problem to the existing getter command parameter block. Users can try to get a non-existant parameter number.
2. What do you mean?

KlemenDEV commented:
> If you are talking about getter procedure blocks, this is a similar problem to the existing getter command parameter block. Users can try to get a non-existant parameter number.

With new improved procedure editor, this should not be that case, as users expect blockly to validate this, so you should make a dependency system

> What do you mean?

Arguments is own tab in editor, but should be on the same page as command name and permissions I would say

Goldorion commented:
> > If you are talking about getter procedure blocks, this is a similar problem to the existing getter command parameter block. Users can try to get a non-existant parameter number.
> 
> With new improved procedure editor, this should not be that case, as users expect blockly to validate this, so you should make a dependency system of some sort for this with validation.
> 
> It is a whole new system that is now targeting more wide user base so it should also be safeguarded as such.
> 
> > What do you mean?
> 
> Arguments is own tab in editor, but should be on the same page as command name and permissions I would say

I put the UI, so the Blockly editor is now on the same page as the property parameters. However, concerning the argument name validator, I really don't know how it could be implemented correctly. I have one idea, but this is not a user-friendly one. My idea would be to add an extension to getter procedure blocks, so users select the related command. Then, using this field, we can check if the command contains the argument name the user wrote inside the getter procedure block. This is the only way I can think to solve the problem, but I know this is not the best one. :/

KlemenDEV commented:
I see it will compile anyways as arguments are strings. But then, at least make all procedures using cmdargs work even if the argument is not passed

KlemenDEV commented on plugins/mcreator-localization/lang/texts.properties:
```diff
@@ -369,6 +371,7 @@ dialog.preferences.page_plugins=Manage plugins
 dialog.preferences.page_themes=UI themes
 dialog.preferences.page_procedure_templates=Procedure templates
 dialog.preferences.page_ai_builder_templates=AI builder templates
+dialog.preferences.page_cmd_builder_templates=Command arguments templates
```
This should be something shorter so the preferences list does not get too wide

Defeatomizer replied:
We can add a disabled **Templates** label to the preferences list before initializing any template panels to indicate that any following tabs offer editing diverse templates, which will enable this string to be shortened to _Command arguments_.

KlemenDEV replied:
It is not the string here, but how it looks in UI and moving location where templates is added does not change that

Goldorion replied:
> This should be something shorter so the preferences list does not get too wide

What do you suggest? We can't write something more precise than `Command arguments`.

KlemenDEV replied:
Cmd. argument templates maybe? Not sure, but right now it is too wide IMO :D

Defeatomizer replied:
Only 4 chars shorter, is that crucial?

KlemenDEV replied:
Open MCreator and you can see it is really wide compared to the whole UI, also if you have idea for shorter variant, it is very welcome, that was just my initial guess

KlemenDEV replied:
Not ideal, but maybe "Command templates" or even "Cmd. templates" could work

Defeatomizer replied:
```suggestion
dialog.preferences.page_cmd_builder_templates=Command templates
```

KlemenDEV commented on src/test/java/net/mcreator/integration/TestWorkspaceDataProvider.java:
```diff
@@ -514,7 +514,7 @@ private static GeneratableElement getExampleFor(ModElement modElement, Random ra
 			command.commandName = modElement.getName();
 			command.permissionLevel = getRandomItem(random, new String[] { "No requirement", "1", "2", "3", "4" });
-			command.onCommandExecuted = new Procedure("procedure2");
+			command.argsxml = "<xml><block type=\"args_start\" deletable=\"!_true\"></block></xml>";
```
As you do separate tests for commands, you can remove command from here entirely, the same as it is with procedure mod element

KlemenDEV commented:
> [](https://github.com/Goldorion)@Goldorion
[Remove commands from tests entirely](https://github.com/MCreator/MCreator/pull/929/commits/834a9277440b6b30ab85aa3a7d0bd700a8fe0207)

Commands should be tested, though, so draft until there are tests for this

Goldorion commented:
> > @Goldorion
> > [Remove commands from tests entirely](https://github.com/MCreator/MCreator/pull/929/commits/834a9277440b6b30ab85aa3a7d0bd700a8fe0207)
> 
> Commands should be tested, though, so draft until there are tests for this

You asked me to remove them, but now you want me to test them? I'm not sure to understand what you want. 🤨

Defeatomizer commented:
I guess Klemen meant you should add them to `GTSampleElements`.

KlemenDEV reviewed (commented):
When converting existing commands, I get this error:

> java.lang.NullPointerException: Cannot invoke "com.google.gson.JsonElement.getAsJsonObject()" because the return value of "com.google.gson.JsonObject.get(String)" is null
	at net.mcreator.element.converter.fv28.CommandArgumentBlockFixer.convert(CommandArgumentBlockFixer.java:34) ~[MCreator.main/:?]
	at net.mcreator.element.GeneratableElement$GSONAdapter.lambda$deserialize$1(GeneratableElement.java:139) ~[MCreator.main/:?]
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183) ~[?:?]
	at java.util.ArrayList.forEach(ArrayList.java:1511) ~[?:?]
	at java.util.stream.SortedOps$RefSortingSink.end(SortedOps.java:395) ~[?:?]
	at java.util.stream.Sink$ChainedReference.end(Sink.java:258) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:510) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499) ~[?:?]
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150) ~[?:?]
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596) ~[?:?]
	at net.mcreator.element.GeneratableElement$GSONAdapter.deserialize(GeneratableElement.java:135) ~[MCreator.main/:?]
	at net.mcreator.element.GeneratableElement$GSONAdapter.deserialize(GeneratableElement.java:90) ~[MCreator.main/:?]
	at com.google.gson.internal.bind.TreeTypeAdapter.read(TreeTypeAdapter.java:69) ~[gson-2.9.0.jar:?]
	at com.google.gson.Gson.fromJson(Gson.java:991) ~[gson-2.9.0.jar:?]
	at com.google.gson.Gson.fromJson(Gson.java:956) ~[gson-2.9.0.jar:?]
	at com.google.gson.Gson.fromJson(Gson.java:905) ~[gson-2.9.0.jar:?]
	at com.google.gson.Gson.fromJson(Gson.java:876) ~[gson-2.9.0.jar:?]
	at net.mcreator.workspace.elements.ModElementManager.fromJSONtoGeneratableElement(ModElementManager.java:113) ~[MCreator.main/:?]
	at net.mcreator.workspace.elements.ModElementManager.loadGeneratableElement(ModElementManager.java:95) ~[MCreator.main/:?]
	at net.mcreator.workspace.elements.ModElement.getGeneratableElement(ModElement.java:94) ~[MCreator.main/:?]
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197) ~[?:?]
	at java.util.stream.SortedOps$SizedRefSortingSink.end(SortedOps.java:357) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:510) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499) ~[?:?]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682) ~[?:?]
	at net.mcreator.generator.Generator.getModBaseGeneratorTemplatesList(Generator.java:397) ~[MCreator.main/:?]
	at net.mcreator.generator.Generator.generateBase(Generator.java:176) ~[MCreator.main/:?]
	at net.mcreator.generator.Generator.generateBase(Generator.java:163) ~[MCreator.main/:?]
	at net.mcreator.generator.Generator.generateBase(Generator.java:153) ~[MCreator.main/:?]
	at net.mcreator.ui.dialogs.workspace.WorkspaceGeneratorSetupDialog.lambda$finalizeTheSetup$2(WorkspaceGeneratorSetupDialog.java:118) ~[MCreator.main/:?]
	at java.lang.Thread.run(Thread.java:833) ~[?:?]

KlemenDEV commented on src/main/java/net/mcreator/ui/modgui/CommandGUI.java:
```diff
@@ -36,35 +45,51 @@
+	private final CompileNotesPanel compileNotesPanel = new CompileNotesPanel();
 	private final ValidationGroup page1group = new ValidationGroup();
+	private BlocklyPanel blocklyPanel;
+	private boolean hasErrors = false;
```
How is this used, it is only set?

Defeatomizer commented on src/main/java/net/mcreator/element/converter/fv26/CommandArgumentBlockFixer.java:
```diff
@@ -31,8 +30,14 @@
-		String procedure = jsonElementInput.getAsJsonObject().get("definition").getAsJsonObject().get("onCommandExecuted").getAsJsonObject().get("name").getAsString();
-		command.argsxml = "<xml><block type=\"args_start\" deletable=\"false\" x=\"40\" y=\"40\"><next><block type=\"call_procedure\"><field name=\"procedure\">" + procedure + "</field></block></next></block></xml>";
+		String procedure = jsonElementInput.getAsJsonObject().get("definition").getAsJsonObject()
+				.get("onCommandExecuted").getAsJsonObject().get("name").getAsString();
```
The `onCommandExecuted` can be null in cases the procedure was not selected (so the default one), I guess this is the reason of conversion problems? Then you should check if the _definition_ JSON object `has("onCommandExecuted")`.

Goldorion replied:
This is surely the reason. I tested multiple times the converter, but I never thought to test it without selecting a procedure.

KlemenDEV reviewed (commented):
Increase workspace FV

Goldorion commented:
> Increase workspace FV

THis is what IU was doing when you were writing your comment ;)

KlemenDEV commented on plugins/mcreator-localization/lang/texts.properties:
```diff
@@ -1757,6 +1773,8 @@ elementgui.block.is_ladder=<html>Does block act like ladder?<br><small>Block mus
 elementgui.block.flammability=Block flammability:
 elementgui.block.push_reaction=Reaction to being pushed:
+elementgui.block.properties_general=General properties
+elementgui.block.properties_dropping=Dropping properties
```
Please remove these texts that don't relate to any of the stuff in this PR, check if there are more like this

Goldorion replied:
Uh... why Intellij? 🤔 😂

Goldorion replied:
I checked and it seems they were the only ones.

KlemenDEV reviewed (commented):
Some validation or input cleanup is needed on name inputs, this for example will fail to build:

![image](https://user-images.githubusercontent.com/16374228/163684596-f3d736d9-aa45-4e8e-987a-4d08c76ce7ec.png)

Defeatomizer commented on plugins/generator-1.16.5/forge-1.16.5/cmdargs/blockpos.java.ftl:
```diff
@@ -1 +1 @@
-.then(Commands.argument("${field$name}", BlockPosArgument.blockPos())${statement$args})
\ No newline at end of file
+.then(Commands.argument("${input$name}", BlockPosArgument.blockPos())${statement$args})
```
Quotations marks why?

MagicalAlexey commented:
cool stuff

KlemenDEV commented on plugins/mcreator-localization/lang/texts.properties:
```diff
@@ -73,6 +73,8 @@ action.gradle.close_mcreator_while_running_title=Close workspace
 dialog.ai_condition.additional_continue=Additional continue condition
 dialog.ai_condition.panel_name=AI condition editor
+dialog.arg_procedure.panel_name=Parameter procedure editor
+dialog.arg_procedure.execute_action=Execute action procedure
```
Where is this key used? Please re-check keys, you did not check them fully yesterday I asked

KlemenDEV commented on src/main/java/net/mcreator/blockly/java/blocks/ProcedureCallBlock.java:
```diff
@@ -81,30 +84,41 @@ public class ProcedureCallBlock implements IBlockGenerator {
+				}
+
+				if (master instanceof BlocklyToJava blocklyToJava
+						&& blocklyToJava.getEditorType() == BlocklyEditorType.COMMAND_ARG) {
```
In case of command arg, you should also check for dependencies to make sure procedure does not request dependences that do not exist

Goldorion replied:
How do I check if the dependencies are correct? All dependencies of the procedure are added to the `BlocklyToCode`, so I can not check the differences between both.

KlemenDEV replied:
Check my commit fixing this, do any changes if you see they are needed :)

Goldorion replied:
I just added the procedure's name causing the problem, so it will be easier to find the source of the problem.

Defeatomizer commented on plugins/generator-1.16.5/forge-1.16.5/cmdargs/_old_command.java.ftl:
```diff
@@ -0,0 +1,23 @@
+		entity = FakePlayerFactory.getMinecraft(world);
+
+	HashMap<String, String> cmdparams = new HashMap<>();
+	int[] index = { -1 };
+	Arrays.stream(cmdargs.getInput().split("\\s+")).forEach(param -> {
+		if(index[0] >= 0)
+			cmdparams.put(Integer.toString(index[0]), param);
+		index[0]++;
+	});
```
I removed this approach from tests in #2584,
>I suggest you do the same.
```suggestion
	int index = -1;
	for (String param : cmdargs.getInput().split("\\s+")) {
		if (index >= 0)
			cmdparams.put(Integer.toString(index), param);
		index++;
	}
```

Defeatomizer commented on plugins/generator-1.18.2/forge-1.18.2/cmdargs/_old_command.java.ftl:
```diff
@@ -0,0 +1,24 @@
+		entity = FakePlayerFactory.getMinecraft(world);
+
+	HashMap<String, String> cmdparams = new HashMap<>();
+    int[] index = { -1 };
+    Arrays.stream(cmdargs.getInput().split("\\s+")).forEach(param -> {
+        if(index[0] >= 0)
+    		cmdparams.put(Integer.toString(index[0]), param);
+    	index[0]++;
+    });
```
```suggestion
	int index = -1;
	for (String param : cmdargs.getInput().split("\\s+")) {
		if (index >= 0)
			cmdparams.put(Integer.toString(index), param);
		index++;
	}
```

Defeatomizer commented on src/main/java/net/mcreator/element/converter/fv29/CommandParameterBlockFixer.java:
```diff
@@ -0,0 +1,56 @@
+			if (!procedure.isEmpty())
+				command.argsxml =
+						"<xml><block type=\"args_start\" deletable=\"false\" x=\"40\" y=\"40\"><next><block type=\"old_command\"><field name=\"procedure\">"
+								+ procedure + "</field></block></next></block></xml>";
```
Add braces here (to both `if` and `else`) so that the following `else` doesn't look lost.

Defeatomizer commented on src/main/java/net/mcreator/element/types/Command.java:
```diff
@@ -18,20 +18,27 @@
 	public String permissionLevel;
 
-	public Procedure onCommandExecuted;
+	public String argsxml;
```
I'd remove empty lines between these 3 fields.

KlemenDEV replied:
They are ok, they split different categories.. it is our convention

Defeatomizer commented on src/main/java/net/mcreator/minecraft/MinecraftImageGenerator.java:
```diff
@@ -554,21 +554,42 @@ public static BufferedImage generateAchievementPreviewPicture(Workspace workspac
 		 * @return <p>Returns generated image.</p>
 		 */
-		public static BufferedImage generateCommandPreviewPicture(String command) {
+		public static BufferedImage generateBasicCommandPreviewPicture(String command, String procedurexml) {
```
Why is this called _basic preview_, is there gonna be some kind of _advanced preview_ later?

Goldorion replied:
The basic preview is the current preview (`/commandName`), while the advanced preview is the procedure style preview.

Defeatomizer commented on src/main/java/net/mcreator/ui/blockly/BlocklyEditorType.java:
```diff
@@ -21,7 +21,8 @@
 
-	PROCEDURE("procedures", "ptpl", "event_trigger"), AI_TASK("ai_setup", "aitpl", "aitasks_container");
+	PROCEDURE("procedures", "ptpl", "event_trigger"), AI_TASK("ai_setup", "aitpl", "aitasks_container"),
+	COMMAND_ARG("cmd_setup", "cmdtpl", "args_start");
```
Maybe move these to separate lines (1 enum constant per line) and disable code formatter on them?

Goldorion replied:
It was Klemen's choice to make them on 1 line.

KlemenDEV replied:
Formatter does it like this

Defeatomizer commented on src/test/java/net/mcreator/integration/TestWorkspaceDataProvider.java:
```diff
@@ -485,13 +485,7 @@ private static GeneratableElement getExampleFor(ModElement modElement, Random ra
-			command.onCommandExecuted = new Procedure("procedure2");
-			return command;
-		} else if (ModElementType.KEYBIND.equals(modElement.getType())) {
+		}else if (ModElementType.KEYBIND.equals(modElement.getType())) {
```
The change there is redundant.
```suggestion
		} else if (ModElementType.KEYBIND.equals(modElement.getType())) {
```

Defeatomizer reviewed (commented):
Idk why I should review this, but here I come to hit 100 comments.

KlemenDEV reviewed (changes requested):
Tests were not tested at all, actions are not even generated... asking once again, test PRs.

Goldorion commented:
> Tests were not tested at all, actions are not even generated... asking once again, test PRs.

When I made the tests for command arguments, I tested them and they worked fine. New tests were executed. The problem happened after and I never noticed it.

KlemenDEV commented:
If you do changes later, you need to test them

KlemenDEV reviewed (changes requested):
@Goldorion, I did more tests and 1.18.2 seems fine, but 1.16.5 gets a build error in some cases. Please see attached workspace and fix this problem, then we can proceed with merging this in the side branch so you can merge args.

Test workspace: [demo.zip](https://github.com/MCreator/MCreator/files/8505515/demo.zip)

Goldorion commented:
> If you do changes later, you need to test them

You also did not test them with your `field_javaname`. Tests were not updated to test `field_javaname` fields.  ;)

KlemenDEV commented on src/test/java/net/mcreator/integration/generator/GTCommandArgBlocks.java:
```diff
@@ -101,7 +101,7 @@ public static void runTest(Logger LOG, String generatorName, Random random, Work
 									processed++;
 									break;
-								case "field_input":
+								case "field_javaname":
```
You should add this as an additional case for field_javaname, keep field_input in case some other block adds it

KlemenDEV replied:
Like in https://github.com/MCreator/MCreator/commit/89c06599c6881eef06d0efef6da3ddbc1f442e28

Goldorion replied:
I was not sure if it would be useless or for this PR.

KlemenDEV replied:
Just to future proof it :)

KlemenDEV commented:
You requested review, does this mean 1.16.5 bug is fixed?

Defeatomizer commented:
Is my review still needed? The PR doesn't need help tips.

KlemenDEV commented:
Testing with the workspace I provided? That is interesting, did you change the template?

Goldorion commented:
> Testing with the workspace I provided? That is interesting, did you change the template?

I tested using your workspace. I opened it, and MCreator set up the workspace and regenerated the code. Build succeeded. I regenerated the code myself. Build succeeded. I opened the command and the name disappeared (I guess this is due to the change of input to field_javaname), so I wrote a name (`name`) and save the mod element. Then, I got the error I fixed. When re-testing, I had no other issues.

KlemenDEV commented:
@Goldorion, please try with this workspace, it still fails: [demo2.zip](https://github.com/MCreator/MCreator/files/8519444/demo2.zip)

MagicalAlexey commented:
I'm waiting for this feature in this version
![image](https://user-images.githubusercontent.com/69478092/164323706-a120ae18-68f4-40fb-a1b8-58a15d647f30.png)

MagicalAlexey commented:
![image](https://user-images.githubusercontent.com/69478092/164324160-3568c0ef-8801-45ea-ac6b-d6352c098bc4.png)

Goldorion commented:
> @Goldorion, please try with this workspace, it still fails: [demo2.zip](https://github.com/MCreator/MCreator/files/8519444/demo2.zip)

I'm not sure why it happened, but the usage of `Collections.EMPTY_MAP` instead of `Collections.emptyMap()`  when calling a procedure without dependencies caused the error.

Defeatomizer commented:
>I'm not sure why it happened, but the usage of `Collections.EMPTY_MAP` instead of `Collections.emptyMap()` when calling a procedure without dependencies caused the error.

I guess that is the normal order of things.
